xla/backends/profiler/gpu/cupti_tracer.cc

```diff
@@ -1184,22 +1235,742 @@ absl::Status CuptiTracer::DisableApiTracing() {
+  for (int devIdx = 0; devIdx < num_gpus_; devIdx++) {
+    // Create a new PM sampling instance for this device
+    std::shared_ptr<PmSamplingDevice> dev = std::make_shared<PmSamplingDevice>(
+        devIdx, option_->pm_sampling_config);
+
+    // Create all configuration needed for this device, or skip device on error
+    if (!dev->CreateConfig().ok()) continue;
+
+    // Set configuration
+    if (!dev->SetConfig().ok()) continue;
+
+    // Device is fully configured but PM sampling not yet started - push to list
+    // of PM sampling devices
+    option_->pm_sampling_config->devices.push_back(std::move(dev));
+  }
+
+  // OK to have no devices that support PM sampling
+  if (option_->pm_sampling_config->devices.size() < 1) return absl::OkStatus();
+
+  // Create decode thread(s)
+  for (int i = 0; i < option_->pm_sampling_config->devices.size(); i +=
+      option_->pm_sampling_config->devs_per_decode_thd) {
+    // Slice iterators
+    auto begin = option_->pm_sampling_config->devices.begin() + i;
+    auto end = begin + option_->pm_sampling_config->devs_per_decode_thd;
+    // Don't go past end of vector
+    end = std::min(end, option_->pm_sampling_config->devices.end());
+
+    // Slice for this decode thread
+    std::vector< std::shared_ptr<PmSamplingDevice> > slice(begin, end);
+
+    // Create worker thread for this slice
+    auto thd = std::make_unique<PmSamplingDecodeThread>(slice);
+    option_->pm_sampling_config->threads.push_back(std::move(thd));
+  }
+
+  // Wait for signal that all threads are ready
+  for (auto& thd : option_->pm_sampling_config->threads) {
+    while (!(thd->IsThdInitialized()));
+  }
+
+  // Start sampling on all devices before enabling decoding, otherwise first
+  // sample contains accumulated results to this point
+  for (auto& dev : option_->pm_sampling_config->devices) {
+    if (auto ret = dev->StartSampling(); !ret.ok()) {
+      LOG(WARNING) << "Profiling::PM Sampling - failed to start on device " <<
+          dev->device_id_;
+      // FIXME: What is appropriate behavior if start thread fails?
+      // Most likely should delete the sampler for this device but this would
+      // need to be communicated to the decoder thread.  Should be safe to do
+      // nothing, as there will just be no data to decode
+    }
+  }
+
+  // Signal threads should be enabled
+  for (auto& thd : option_->pm_sampling_config->threads) {
+    thd->EnableThd();
+  }
+
+  // Wait for signal that decode thread is enabled
+  for (auto& thd : option_->pm_sampling_config->threads) {
+    while (!(thd->IsThdEnabled()));
+  }
+
+  pm_sampling_enabled_ = true;
+
+  return absl::OkStatus();
+}
+
+absl::Status CuptiTracer::DisablePMSampling() {
+  ERROR_IF_PM_SAMPLING_NOT_SUPPORTED();
+  if (!pm_sampling_enabled_) return absl::OkStatus();
+
+  // Stop sampling on all devices
+  for (auto& dev : option_->pm_sampling_config->devices) {
+    if (auto ret = dev->StopSampling(); !ret.ok()) {
```

Comment:
Log the error status, other you can just write:

```suggestion
    if (!dev-> StopSampling().ok()) {
```